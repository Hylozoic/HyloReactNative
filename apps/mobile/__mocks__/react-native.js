import React, {useState} from 'react'
import * as ReactNative from 'react-native'

export const Switch = props => {
  const [value, setValue] = useState(props.value)
  return (
    <ReactNative.TouchableOpacity
      onPress={() => {
        props.onValueChange(!value)
        setValue(!value)
      }}
      testID={props.testID}>
      <ReactNative.Text>{value?.toString()}</ReactNative.Text>
    </ReactNative.TouchableOpacity>
  )
}

newRN = Object.defineProperty(ReactNative, 'Switch', {
  get: function() {
    return Switch
  }
})

module.exports = newRN
