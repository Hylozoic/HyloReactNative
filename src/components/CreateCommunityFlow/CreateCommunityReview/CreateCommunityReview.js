import React from 'react'
import {
  Text,
  View,
  TextInput,
  TouchableOpacity
} from 'react-native'
import Button from '../../Button'
import KeyboardFriendlyView from '../../KeyboardFriendlyView'
import ErrorBubble from '../../ErrorBubble'
import {
  formatDomainWithUrl
} from '../util'
import createCommunityHeader from '../createCommunityHeader'
import styles from '../CreateCommunityFlow.styles'

export default class CreateCommunityReview extends React.Component {
  static navigationOptions = ({ navigation }) => {
    const title = 'STEP 3/3'
    return createCommunityHeader(title, navigation)
  }

  constructor (props) {
    super(props)
    this.state = {
      communityName: this.props.communityName,
      communityUrl: this.props.communityUrl,
      error: null
    }
  }

  submit = () => {
    const { communityName, communityUrl } = this.state
    this.props.createCommunity(communityName, communityUrl)
    .then((data) => {
      if (data.error) {
        this.setState({
          error: 'There was an error, please try again.'
        })
      }
      this.props.clearNameAndUrlFromStore()
      this.props.goToCommunity(data.payload.data.createCommunity.community)
    })
  }

  render () {
    const { error } = this.state
    return <KeyboardFriendlyView style={styles.container}>
      <Text style={styles.header}>Everything look good?</Text>
      <Text style={styles.description}>You can always come back and change your details at any time</Text>
      <View style={styles.reviewTextInputContainer}>
        <Text style={styles.textInputLabel}>Whats the name of your community?</Text>
        <View style={styles.textInputWithButton}>
          <TextInput
            style={styles.reviewTextInput}
            onChangeText={communityName => this.setInput('communityName', communityName)}
            returnKeyType='next'
            autoCapitalize='none'
            value={this.state.communityName}
            autoCorrect={false}
            underlineColorAndroid={styles.androidInvisibleUnderline}
            disabled
          />
          <TouchableOpacity onPress={this.props.goToCreateCommunityName} style={styles.editContainer}>
            <Text style={styles.editText}>Edit</Text>
          </TouchableOpacity>
        </View>
      </View>
      <View style={styles.secondTextInputContainer}>
        <Text style={styles.textInputLabel}>Whats the url of your community?</Text>
        <View style={styles.textInputWithButton}>
          <TextInput
            style={styles.reviewTextInput}
            onChangeText={communityName => this.setInput('communityName', communityName)}
            returnKeyType='next'
            autoCapitalize='none'
            value={formatDomainWithUrl(this.state.communityUrl)}
            autoCorrect={false}
            underlineColorAndroid={styles.androidInvisibleUnderline}
            disabled
          />
          <TouchableOpacity onPress={this.props.goToCreateCommunityUrl} style={styles.editContainer}>
            <Text style={styles.editText}>Edit</Text>
          </TouchableOpacity>
        </View>
      </View>
      {error && <View style={styles.errorBubble}><ErrorBubble text={error} /></View> }
      <Button text="Let's Do This!" onPress={this.submit} style={styles.button} disabled={!!this.props.createCommunityPending} />
    </KeyboardFriendlyView>
  }
}
