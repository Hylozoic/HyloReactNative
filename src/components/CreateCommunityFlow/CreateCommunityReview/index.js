import component from './CreateCommunityReview'
export default component
