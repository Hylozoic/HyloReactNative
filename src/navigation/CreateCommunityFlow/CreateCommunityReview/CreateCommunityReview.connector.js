import { connect } from 'react-redux'
import selectCommunity from 'store/actions/selectCommunity'
import {
  createCommunity,
  clearNameAndUrlFromStore,
  CREATE_COMMUNITY,
  getCommunityName,
  getCommunityUrl
} from '../CreateCommunityFlow.store'

export function mapStateToProps (state, props) {
  const communityName = getCommunityName(state)
  const communityUrl = getCommunityUrl(state)
  const createCommunityPending = state.pending[CREATE_COMMUNITY]
  return {
    communityName,
    communityUrl,
    createCommunityPending
  }
}

export function mapDispatchToProps (dispatch, props) {
  return {
    createCommunity: (name, slug) => dispatch(createCommunity(name, slug)),
    clearNameAndUrlFromStore: () => dispatch(clearNameAndUrlFromStore()),
    selectCommunity: (id) => dispatch(selectCommunity(id))
  }
}

export function mergeProps (stateProps, dispatchProps, ownProps) {
  const { navigation } = ownProps
  return {
    ...stateProps,
    ...dispatchProps,
    ...ownProps,
    goToCreateCommunityName: () => {
      navigation.navigate('CreateCommunityName')
    },
    goToCreateCommunityUrl: () => {
      navigation.navigate('CreateCommunityUrl')
    },
    goToCommunity: community => {
      navigation.closeDrawer()
      const { id } = community
      dispatchProps.selectCommunity(id)
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)
