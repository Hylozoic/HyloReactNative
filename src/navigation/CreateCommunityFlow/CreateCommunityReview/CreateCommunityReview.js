import React from 'react'
import {
  Text,
  View,
  TextInput,
  TouchableOpacity
} from 'react-native'
import Button from 'components/Button'
import KeyboardFriendlyView from 'navigation/KeyboardFriendlyView'
import { keyboardAvoidingViewProps as kavProps } from 'util/viewHelpers'
import ErrorBubble from 'components/ErrorBubble'
import styles from '../CreateCommunityFlow.styles'
import SafeAreaView from 'react-native-safe-area-view'

export default class CreateCommunityReview extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      communityName: this.props.communityName,
      communityUrl: this.props.communityUrl,
      error: null
    }
  }

  submit = () => {
    const { communityName, communityUrl } = this.state
    const { createCommunity, clearNameAndUrlFromStore, goToCommunity } = this.props
    return createCommunity(communityName, communityUrl)
      .then((data) => {
        if (data.error) {
          this.setState({
            error: 'There was an error, please try again.'
          })
          return
        }
        clearNameAndUrlFromStore()
        goToCommunity(data.payload.data.createCommunity.community)
      })
      .catch(() => {
        this.setState({
          error: 'There was an error, please try again.'
        })
      })
  }

  render () {
    const { error } = this.state
    return (
      <SafeAreaView style={styles.container}>
        <KeyboardFriendlyView {...kavProps}>
          <View style={styles.header}>
            <Text style={styles.heading}>Everything look good?</Text>
            <Text style={styles.description}>You can always come back and change your details at any time</Text>
          </View>
          <View style={styles.content}>
            <View style={styles.textInputContainer}>
              <Text style={styles.textInputLabel}>What's the name of your community?</Text>
              <View style={styles.textInputWithButton}>
                <TextInput
                  style={styles.textInput}
                  value={this.state.communityName}
                  underlineColorAndroid={styles.androidInvisibleUnderline}
                  disabled
                />
                <TouchableOpacity onPress={this.props.goToCreateCommunityName} style={styles.editContainer}>
                  <Text style={styles.editText}>Edit</Text>
                </TouchableOpacity>
              </View>
            </View>
            <View style={styles.textInputContainer}>
              <Text style={styles.textInputLabel}>What's the url of your community?</Text>
              <View style={styles.textInputWithButton}>
                <TextInput
                  style={styles.textInput}
                  value={this.state.communityUrl}
                  underlineColorAndroid={styles.androidInvisibleUnderline}
                  disabled
                />
                <TouchableOpacity onPress={this.props.goToCreateCommunityUrl} style={styles.editContainer}>
                  <Text style={styles.editText}>Edit</Text>
                </TouchableOpacity>
              </View>
            </View>
            {error && <View style={styles.errorBubble}><ErrorBubble text={error} /></View>}
          </View>
          <View style={styles.footer}>
            <Button
              text="Let's Do This!"
              onPress={this.submit}
              style={styles.button}
              disabled={!!this.props.createCommunityPending}
            />
          </View>
        </KeyboardFriendlyView>
      </SafeAreaView>
    )
  }
}
