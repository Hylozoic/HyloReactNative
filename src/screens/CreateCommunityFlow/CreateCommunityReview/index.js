import component from './CreateCommunityReview'
import connector from './CreateCommunityReview.connector'
export default connector(component)
