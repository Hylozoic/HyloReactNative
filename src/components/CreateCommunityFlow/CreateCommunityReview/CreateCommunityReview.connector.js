import { connect } from 'react-redux'
import selectCommunity from '../../../store/actions/selectCommunity'
import { createCommunity, clearNameAndUrlFromStore } from '../CreateCommunityFlow.store'
import { ALL_COMMUNITIES_ID } from '../../../store/models/Community'

export function mapStateToProps (state, props) {
  const { communityName, communityUrl } = state.CreateCommunityFlow
  return {
    communityName,
    communityUrl
  }
}

export function mapDispatchToProps (dispatch, props) {
  return {
    createCommunity: (name, slug) => dispatch(createCommunity(name, slug)),
    clearNameAndUrlFromStore: () => dispatch(clearNameAndUrlFromStore()),
    selectCommunity: (id) => dispatch(selectCommunity(id))
  }
}

export function mergeProps (stateProps, dispatchProps, ownProps) {
  const { navigation } = ownProps
  return {
    ...stateProps,
    ...dispatchProps,
    ...ownProps,
    goToCreateCommunityName: () => {
      navigation.navigate('CreateCommunityName')
    },
    goToCreateCommunityUrl: () => {
      navigation.navigate('CreateCommunityUrl')
    },
    goToCommunity: community => {
      if (community.id === ALL_COMMUNITIES_ID &&
        ownProps.screenProps.currentTabName !== 'Home') {
        navigation.navigate('Home')
      } else {
        navigation.navigate('DrawerClose')
      }
      const { id } = community
      dispatchProps.selectCommunity(id)
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)
