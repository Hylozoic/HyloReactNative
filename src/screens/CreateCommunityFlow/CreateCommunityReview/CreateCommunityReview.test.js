import 'react-native'
import React from 'react'
import ReactShallowRenderer from 'react-test-renderer/shallow'
import ReactTestRenderer from 'react-test-renderer'
import CreateCommunityReview from './CreateCommunityReview'

jest.mock('components/KeyboardFriendlyView', () => 'KeyboardFriendlyView')

it('matches last snapshot', () => {
  const renderer = new ReactShallowRenderer()

  renderer.render(<CreateCommunityReview />)
  const actual = renderer.getRenderOutput()
  expect(actual).toMatchSnapshot()
})

it('matches last snapshot when visible is true', () => {
  const renderer = ReactTestRenderer.create(<CreateCommunityReview />)
  const instance = renderer.getInstance()
  instance.setState({ error: 'This is an error' })
  expect(renderer.toJSON()).toMatchSnapshot()
})

it('calls createCommunity on submit', () => {
  const result = {
    payload: {
      data: {
        createCommunity: {
          community: {
            id: 1
          }
        }
      }
    }
  }

  const props = {
    createCommunity: jest.fn(() => Promise.resolve(result)),
    clearNameAndUrlFromStore: jest.fn(),
    goToCommunity: jest.fn()
  }
  const communityName = 'communityName'
  const communityUrl = 'communityUrl'
  const renderer = ReactTestRenderer.create(<CreateCommunityReview {...props} />)
  const instance = renderer.getInstance()
  instance.setState({
    communityName,
    communityUrl
  })
  return instance.submit()
    .then(() => {
      expect(props.createCommunity).toHaveBeenCalledWith(communityName, communityUrl)
      expect(props.clearNameAndUrlFromStore).toHaveBeenCalled()
      expect(props.goToCommunity).toHaveBeenCalledWith(result.payload.data.createCommunity.community)
    })
})

it('creates an error on submit if an error is returned', () => {
  const result = {
    error: 'some error'
  }
  const props = {
    createCommunity: jest.fn(() => Promise.resolve(result)),
    clearNameAndUrlFromStore: jest.fn(),
    goToCommunity: jest.fn()
  }
  const communityName = 'communityName'
  const communityUrl = 'communityUrl'
  const renderer = ReactTestRenderer.create(<CreateCommunityReview {...props} />)
  const instance = renderer.getInstance()
  instance.setState({
    communityName,
    communityUrl
  })
  return instance.submit()
    .then(() => {
      expect(instance.state.error).toBeTruthy()
    })
})

it('creates an error on submit if the promise is rejected', () => {
  const props = {
    createCommunity: jest.fn(() => Promise.reject(new Error(''))),
    clearNameAndUrlFromStore: jest.fn(),
    goToCommunity: jest.fn()
  }
  const communityName = 'communityName'
  const communityUrl = 'communityUrl'
  const renderer = ReactTestRenderer.create(<CreateCommunityReview {...props} />)
  const instance = renderer.getInstance()
  instance.setState({
    communityName,
    communityUrl
  })
  return instance.submit()
    .then(() => {
      expect(instance.state.error).toBeTruthy()
    })
})
