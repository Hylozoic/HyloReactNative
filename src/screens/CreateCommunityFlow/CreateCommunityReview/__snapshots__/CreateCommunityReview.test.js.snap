// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches last snapshot 1`] = `
<SafeAreaViewMock>
  <KeyboardFriendlyView>
    <View>
      <Text>
        Everything look good?
      </Text>
      <Text>
        You can always come back and change your details at any time
      </Text>
    </View>
    <View>
      <View>
        <Text>
          What's the name of your community?
        </Text>
        <View>
          <Component
            allowFontScaling={true}
            disabled={true}
            rejectResponderTermination={true}
            underlineColorAndroid="rgba(0,0,0,0)"
          />
          <ForwardRef>
            <Text>
              Edit
            </Text>
          </ForwardRef>
        </View>
      </View>
      <View>
        <Text>
          What's the url of your community?
        </Text>
        <View>
          <Component
            allowFontScaling={true}
            disabled={true}
            rejectResponderTermination={true}
            underlineColorAndroid="rgba(0,0,0,0)"
          />
          <ForwardRef>
            <Text>
              Edit
            </Text>
          </ForwardRef>
        </View>
      </View>
    </View>
    <View>
      <Button
        disabled={false}
        onPress={[Function]}
        text="Let's Do This!"
      />
    </View>
  </KeyboardFriendlyView>
</SafeAreaViewMock>
`;

exports[`matches last snapshot when visible is true 1`] = `
<SafeAreaView>
  <KeyboardFriendlyView>
    <View>
      <Text>
        Everything look good?
      </Text>
      <Text>
        You can always come back and change your details at any time
      </Text>
    </View>
    <View>
      <View>
        <Text>
          What's the name of your community?
        </Text>
        <View>
          <TextInput
            allowFontScaling={true}
            disabled={true}
            rejectResponderTermination={true}
            underlineColorAndroid="rgba(0,0,0,0)"
          />
          <View
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <Text>
              Edit
            </Text>
          </View>
        </View>
      </View>
      <View>
        <Text>
          What's the url of your community?
        </Text>
        <View>
          <TextInput
            allowFontScaling={true}
            disabled={true}
            rejectResponderTermination={true}
            underlineColorAndroid="rgba(0,0,0,0)"
          />
          <View
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <Text>
              Edit
            </Text>
          </View>
        </View>
      </View>
      <View>
        <View>
          <View>
            <Text>
              This is an error
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View>
      <View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <View>
            <View>
              <Text>
                Let's Do This!
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </KeyboardFriendlyView>
</SafeAreaView>
`;
